source/extensions/transport_sockets/http_11_proxy/connect.cc

```diff
@@ -0,0 +1,140 @@
+#include "source/extensions/transport_sockets/http_11_proxy/connect.h"
+
+#include <sstream>
+
+#include "envoy/network/transport_socket.h"
+
+#include "source/common/buffer/buffer_impl.h"
+#include "source/common/common/scalar_to_byte_vector.h"
+#include "source/common/common/utility.h"
+#include "source/common/network/address_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace Http11Connect {
+
+const std::string& UpstreamHttp11ConnectSocket::key() {
+  CONSTRUCT_ON_FIRST_USE(std::string, "envoy.network.transport_socket.http_11_proxy.address");
+}
+
+UpstreamHttp11ConnectSocket::UpstreamHttp11ConnectSocket(
+    Network::TransportSocketPtr&& transport_socket,
+    Network::TransportSocketOptionsConstSharedPtr options)
+    : PassthroughSocket(std::move(transport_socket)), options_(options) {
+  if (options_ && options_->proxyInfo() && transport_socket_->ssl()) {
+    header_buffer_.add(
+        absl::StrCat("CONNECT ", options_->proxyInfo()->hostname, ":443 HTTP/1.1\r\n\r\n"));
+  } else {
+    stripped_connect_ = true;
+  }
+}
+
+void UpstreamHttp11ConnectSocket::setTransportSocketCallbacks(
+    Network::TransportSocketCallbacks& callbacks) {
+  transport_socket_->setTransportSocketCallbacks(callbacks);
+  callbacks_ = &callbacks;
+}
+
+Network::IoResult UpstreamHttp11ConnectSocket::doWrite(Buffer::Instance& buffer, bool end_stream) {
+  if (header_buffer_.length() > 0) {
+    auto header_res = writeHeader();
+    return header_res;
+  } else if (stripped_connect_) {
+    // Don't pass events up until the connect response is read because TLS reads
+    // kick off writes which don't pass through the transport socket.
+    return transport_socket_->doWrite(buffer, end_stream);
+  }
+  return Network::IoResult{Network::PostIoAction::KeepOpen, 0, false};
+}
+
+Network::IoResult UpstreamHttp11ConnectSocket::doRead(Buffer::Instance& buffer) {
+  if (!stripped_connect_) {
+    char peek_buf[200];
+    auto result = callbacks_->ioHandle().recv(peek_buf, 200, MSG_PEEK);
+    if (!result.ok() && result.err_->getErrorCode() != Api::IoError::IoErrorCode::Again) {
+      return {Network::PostIoAction::Close, 0, false};
+    }
+    absl::string_view peek_data(peek_buf, result.return_value_);
+    size_t index = peek_data.find("\r\n\r\n");
+    if (index == absl::string_view::npos) {
+      if (result.return_value_ == 200) {
+        ENVOY_CONN_LOG(trace, "failed to receive CONNECT headers within 200 bytes",
+                       callbacks_->connection());
+        return {Network::PostIoAction::Close, 0, false};
+      }
+      return Network::IoResult{Network::PostIoAction::KeepOpen, 0, false};
```

Comment:
nit: More early return opportunities here :) (Just remove the `else {}` below)

Reply:
done